[aes/dv] Add GCM-specific FI tests, add missing FI SEC_CM labels and testplan entries

vogelpi opened the pull request.

martin-velay committed:
[dv] Fix DVSim after GUI_DEBUG addition

- GUI_DEBUG mode addition is creating an issue when tests are going
into timeout. Because Deploy.py is managing GUI_DEBUG mode through
GUI mode, but in LocalLauncher.py we were trying to access gui_debug
attribute which doesn't exists.

Signed-off-by: Martin Velay <mvelay@lowrisc.org>

vogelpi committed:
[aes/dv] Also fault GHASH-specific signal in main control FSM

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi committed:
[aes/dv] Add GCM.CONFIG.SHADOW SEC_CM label and testplan entry

This countermeasures is already tested as part of the fully automated
CSR tests.

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi committed:
[aes/dv] Test GHASH.FSM.SPARSE countermeasure, add required DV interface

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi committed:
[aes/dv] Test GHASH.FSM.LOCAL_ESC countermeasure

In line with other FI countermeasure tests, the newly added test injects
faults into the GHASH module and verifies that the module locks up and
triggers a fatal alert.

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi committed:
[aes/dv] Test GCM.CONFIG.SPARSE countermeasure

This commit enables testing the GCM.CONFIG.SPARSE in the following ways:

1. The aes_core_fi test is extended to inject invalid values into the
   GCM phase value presented to the internal hardware and ensures that
   with such an invalid value present, the DUT doesn't progress and
   remains idle.

2. The base sequence is extended to randomly write values greater than
   16 or values equal to 0 to the num_valid_bytes field of the GCM
   control register. The hardware resolves such values to 16 and
   continues. Failures for doing so are detected by mismatching tag
   comparisons.

3. The base sequence is extended to all randomly injecting GCM phase
   transitions which are not allowed. Most of these transitions are now
   verified in the aes_config_error test.

4. The aes_gcm_save_restore sequence is extended to randomly inject
   phase transitions which not allowed but which are hard to handle
   in the scoreboard. Also this sequence now performs writes with
   illegal phase values and verifies that the DUT moves back to the
   GCM_INIT phase when receiving such writes.

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi committed:
[aes/dv] Switch erroneously inserted set() call to predict()

This is a follow-up of 1a6506d637ccb779e8365400b9576dcdc26df0a9.

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi committed:
[aes/dv] Drop unnecessary writes to the GCM control CSR in the S&R vseq

The sequence erronously cleared the wrong flag upon writing the GCM
control register during the GCM_TEXT phase.

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

nasahlpa commented on hw/ip/aes/data/aes_sec_cm_testplan.hjson:
```diff
@@ -209,6 +209,15 @@
+      name: sec_cm_ghash_fsm_sparse
+      desc: '''
+            Verify the countermeasure(s) GHASH.FSM.SPARSE.
+            It is ensured that upon randomly forcing bits in the state register of the GHASH FSM, the DUT stops processing data, locks up and triggers a fatal alert.
```
This test plan entry corresponds to the changes in the `aes_fi_vseq`?
If yes, there is the following comment in the code:
`// For the GHASH FSM, we have to force an invalid value to trigger an alert`
Maybe we should update the test plan entry accordingly? E.g.:
```
It is ensured that upon manipulating the state register of the GHASH FSM to an invalid value, the DUT stops processing data, locks up and triggers a fatal alert.
```
Once/if we  have the full protection for the FSM, we again can update this comment?

vogelpi replied:
Yes, this makes sense. I've now corrected this. Thanks for your review.

nasahlpa reviewed (approved):
Thanks for working on this @vogelpi!

vogelpi committed:
[aes/dv] Don't expect to leave GCM_INIT in aes_shadow_reg_errors tests

These tests randomly write the shadow registers inside AES. Until now,
they did not account for the fact that to actually leave the GCM_INIT
phase, the initialization phase needs to complete which is impossible
to hit in these tests. This commit thus aligns the prediction such that
leaving GCM_INIT is not possible in these tests.

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi committed:
[aes/dv] Read back shadowed main control register in aes_init()

It turned out that under some circumstances the preceeding write fails
to update the mirrored values leading to issues in some directed tests
such as aes_man_cfg_err.

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi committed:
[aes/dv] Simplify aes_gcm_len_cg definition

We primarily care about seeing messages with 0, 1, 2 or many AAD/TEXT
blocks and where the last AAD/TEXT block is either full or partial.

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi committed:
[aes/dv] Modify the distribution of the message/AAD length

This commit changes the distributions such that half of the messages
have a full last AAD or TEXT block.

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi committed:
[aes/dv] Add coverage exclusions for GHASH FSM

Some FSM states and transitions are only possible in the masked or
unmasked implementation, respectively. This commit adds the
corresponding coverage exclusions.

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi committed:
[aes] Add SEC_CM label and SVAs for the GHASH masking

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>
